PR2/3 feat(user): Add points awarding logic on challenge completion (#764)

aurelien-darbellay opened the pull request:
# Pull Request Description

### 🎯 Prerequisite and observations
This task depends on the completion of **task 763** (PR #988).

Also, it will generate conflict with PR #983, which changes the enum ChallengeStatus. The recommended workflow is: (1) merge #983 into develop, (2) then rebase the branch "feature/764_add_method_awardsPointsForSolvedChallenge_to_UserSolutionServiceImpl", (3) then adapt to the new instances of ChallengeStatus (ChallengeStatus.ENDED --> ChallengeStatus.SUBMITTED_COMPLETE). 

---

### 🔧 Scope of Work
This PR introduces logic to award points to a user when they complete a challenge, by extending the `UserSolutionService` implementation.

Key changes include:

1. **Dependency Injection**
   - Added `UserService` as a dependency in `UserSolutionServiceImpl` to allow updating user points.

2. **Constant for Awarded Points**
   - Defined a constant in `UserSolutionServiceImpl` to store the number of points that should be awarded upon challenge completion.

3. **New Method: `awardsPointsForSolvedChallenge`**
   - Checks if a `UserSolutionDocument` represents a completed challenge (`status == ENDED`).
   - If completed, calls `userService.addPointsToUser` with the `userId` and the constant for awarded points.
   - Returns a `Mono<UserSolutionDocument>`, making it suitable for use in the `addSolution` reactive pipeline.

4. **Unit Tests**
   - Extended `UserSolutionServiceImplTest` to cover both cases:
     - Solution not completed → returns the document unchanged, no interaction with `userService`.
     - Solution completed → calls `userService.addPointsToUser`, returns the document unchanged.
   - Ensured both branches of logic are verified with `StepVerifier`.

---

### Files modified

- UserSolutionServiceImpl
- UserSolutionServiceImplTest
- Changelog
- .env.CI.dev

---

### ✅ Benefits
- Automates user point awarding upon challenge completion.
- Keeps business rules consistent by storing the awarded points in a single constant.
- Ensures reactive flow remains non-blocking and integrates seamlessly with existing pipelines.
- Strengthens test coverage for new logic, keeping SonarQube happy.

---

### ✅ Version and Changelog
- This PR represents a patch to the app, an internal modification with no behaviour change. Accordingly, the changelog has been updated from 2.2.0 to 2.2.1. 
---

### ✅ Author Self-check

- [x] I tested my changes locally and verified they work as expected  
- [x] The PR has a clear and focused purpose (only one feature, fix or refactor)  
- [x] Title, description, and changelog (if needed) are filled in correctly  
- [x] There are no leftover merge conflicts or unrelated changes from previous branches  
- [x] All automated checks (like SonarCloud) have passed  
- [x] I responded to all previous review comments and only resolved those I truly addressed  
- [x] I ran SonarLint locally on the modified files and confirmed there are no warnings or errors

ToniR90 committed:
feat(UserService): add 'addPointsToUser' to UserService

ToniR90 committed:
feat(UserServiceImpl): added 'addPoints' and 'addPointsToUser' methods to UserServiceImpl

ToniR90 committed:
feat(UserServiceImplTest): added new test for the 'addPoints' method

ToniR90 committed:
update(changelog): updated changelod due to rebase

# Conflicts:
#	CHANGELOG.md

aurelien-darbellay committed:
added feature to UserSolutionServiceImpl: added constant POINT_PER_SOLVED_CHALLENGE; added dependency UserService; created method awardsPointsPerSolvedChallenge

aurelien-darbellay committed:
added tests for method awardsPointsForSolvedChallenge

aurelien-darbellay committed:
changelog updated: itachallenge-user version 2.2.0

# Conflicts:
#	CHANGELOG.md

# Conflicts:
#	CHANGELOG.md

rioacosta reviewed (changes requested):
It's all rigth to me, the only issue is semantic, given you're using ENDED instead of SUBMITTED_COMPLETE as I'm being renaming it in PR's #983 #984 #985, If mines get merged first, we should update this method to reflect the new naming for consistency, if not it'll have to be done sooner than later even though the method hasn't been called yet, but I really don't know what to suggest...

Maybe @ivilarop can enlighten with what to do in this situation!

aurelien-darbellay commented:
@rioacosta 
I've added this to the PR description. 


Also, it will generate conflict with PR https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/983, which changes the enum ChallengeStatus. The recommended workflow is: (1) merge https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/983 into develop, (2) then rebase the branch "feature/764_add_method_awardsPointsForSolvedChallenge_to_UserSolutionServiceImpl", (3) then adapat to the new instances of ChallengeStatus (ChallengeStatus.ENDED --> ChallengeStatus.SUBMITTED_COMPLETE).

MarTorrijos reviewed (commented):
Just a nitpick question. Shouldn't the Changelog be 2.1.0-RELEASE? Right now you have it like 2.2, but I see the previous one was 2.0.

Aside from that I think everything's good! Good explanation of the changes made and new tests were created too 👍

aurelien-darbellay commented:
@MarTorrijos : 2.1.0 is the version resulting from Toni's pull request, on which this one is based. 2.0.0 is the last version that was merged into dev. I don't really know what is the correct practice in those cases. But I guess I could try and rebase on Tonis..

MarTorrijos reviewed (approved):
Aah okay, makes sense. Maybe this is something we should ask Ivan then. 

Aside from that all good

aurelien-darbellay committed:
fix: restored version bump 2.1.0 from task 763 and PR 988 in changelog

aurelien-darbellay committed:
added loggers to awardsPointsForSolvedChallenge method

aurelien-darbellay committed:
added tests for loggers in awardsPointsFOrSolvedChallenge

Meriekje reviewed (changes requested):
Introducing the @Builder annotation is a great idea (and it will make it easier for me as well so thanks). However, as I understand it, using @Builder.Default makes the points field never null already but later in UserServiceImpl you have a safety check "Optional.ofNullable(user.getPoints()).orElse(0)". Is it necessary?
Also, in UserSolutionServiceImpl, you're injecting UserServiceImpl instead of the interface. Wouldn't the opposite be better to test?
By the way, line 86: Wouldn't "status == ChallengeStatus.ENDED" be preferable?

On a minor point, Since this PR includes a changelog in its files changed, it is required to provide a reference in the PR description. Also specify your info from the canghelog to make it easier to identify what's yours since there are so many changes (from other PR).

Lastly, I am not sure why you added the package here: 
Before: public class UserServiceImpl implements UserService {
After: public class UserServiceImpl implements com.itachallenge.user.service.UserService {

Everything else looks pretty good to me!

aurelien-darbellay committed:
inject interface in constructor, instead of implementation

aurelien-darbellay commented:
@Meriekje 
Thanks for the feedback!
(1) I've updated the descripiton to reflect version update
(2) I've changed the dependency injection in UserSolutionServiceImpl... thanks a lot for that. 
(3) Line 86 of UserSolutionServiceImpl doesn't belong to the scope of this PR. 
(4) Everymodification in UserServiceImpl and UserServiceImplTest has to do with Toni's previous PR, not this one. @ToniR90 look at @Meriekje's comment.

aurelien-darbellay committed:
UserServiceImpl implements com.itachallenge.user.service.UserService --> UserServiceImpl implements UserService

aurelien-darbellay committed:
fix:remove typo in import in UserServiceImpl

aurelien-darbellay commented:
@Meriekje I've changed the weird whole package implementation indication. The null safety check, we'd like to keep it - even though UserDocuments will be initialized with 0 by default, we believe that, in the future, there might situation in which the method receives them with a null value.

aurelien-darbellay committed:
.env.CI.dev updated: MICROSERVICE_VERSION_itachallenge-user=2.2.0-RELEASE

aurelien-darbellay committed:
fix: UserServiceImpl implements UserService, not the class referred to with the whole package path

Meriekje commented:
> @Meriekje Thanks for the feedback! (1) I've updated the descripiton to reflect version update (2) I've changed the dependency injection in UserSolutionServiceImpl... thanks a lot for that. (3) Line 86 of UserSolutionServiceImpl doesn't belong to the scope of this PR. (4) Everymodification in UserServiceImpl and UserServiceImplTest has to do with Toni's previous PR, not this one. @ToniR90 look at @Meriekje's comment.

I can see it. However, check the env.CI.dev as it still reflects the old version.
Sorry for addressing you a change from Toni and now I see that line 86 is out of this scope.

ToniR90 committed:
Update .env.CI.dev

Updated .env.CI.dev file to match with the changelog.md version

Meriekje commented:
> @Meriekje I've changed the weird whole package implementation indication. The null safety check, we'd like to keep it - even though UserDocuments will be initialized with 0 by default, we believe that, in the future, there might situation in which the method receives them with a null value.

Ok. Add this info to the PR so everyone knows why it's there

aurelien-darbellay committed:
fix: UserServiceImpl implements UserService, not the class referred to with the whole package path in test too

aurelien-darbellay committed:
Fix: CHANGELOG.md - new version is a patch with no behaviour change

aurelien-darbellay committed:
Fix: .env.Cl.dev - new version is a patch with no behaviour change

aurelien-darbellay committed:
fix: removed phantom changes form CHANGELOG.md

aurelien-darbellay committed:
fix: remove phantom changes from UserServiceImplTest.java

aurelien-darbellay committed:
fix: remove phantom changes UserServiceImplTest.java

aurelien-darbellay committed:
fix: remove phantom change UserService

aurelien-darbellay committed:
fix: removed last phantom changes UserServiceImplTest.java

Meriekje reviewed (approved):
Changes habe been submited so well done!

aurelien-darbellay committed:
fix: remove phantom changes in UserServiceImpl

ivilarop commented:
We are closing the PR because in Sprint 15 we are analyzing the refactoring of the logic and data modeling related to points within the Challenge microservice.
